File: internal/tui/components/keybinds.go
package components

import (
	"strings"

	"github.com/charmbracelet/lipgloss"

	"ash/internal/tui"
)

// KeyBind represents a single key binding.
type KeyBind struct {
	Key         string
	Description string
}

// KeyBinds renders a help footer with key bindings.
type KeyBinds struct {
	styles   tui.Styles
	bindings []KeyBind
	width    int
}

// NewKeyBinds creates a new key bindings component.
func NewKeyBinds(styles tui.Styles) *KeyBinds {
	return &KeyBinds{
		styles: styles,
	}
}

// SetBindings sets the key bindings to display.
func (k *KeyBinds) SetBindings(bindings []KeyBind) {
	k.bindings = bindings
}

// SetWidth sets the component width.
func (k *KeyBinds) SetWidth(width int) {
	k.width = width
}

// Render renders the key bindings.
func (k *KeyBinds) Render() string {
	if len(k.bindings) == 0 {
		return ""
	}

	var parts []string
	for _, bind := range k.bindings {
		key := k.styles.KeyBind.Render(bind.Key)
		desc := k.styles.KeyBindDesc.Render(bind.Description)
		parts = append(parts, key+" "+desc)
	}

	return strings.Join(parts, "  ")
}

// RenderCompact renders key bindings in a compact format.
func (k *KeyBinds) RenderCompact() string {
	if len(k.bindings) == 0 {
		return ""
	}

	var parts []string
	for _, bind := range k.bindings {
		parts = append(parts, bind.Key+":"+bind.Description)
	}

	return k.styles.StatusBar.Render(strings.Join(parts, " | "))
}

// RenderVertical renders key bindings vertically.
func (k *KeyBinds) RenderVertical() string {
	if len(k.bindings) == 0 {
		return ""
	}

	var b strings.Builder
	for _, bind := range k.bindings {
		key := k.styles.KeyBind.Render(bind.Key)
		desc := k.styles.KeyBindDesc.Render(bind.Description)
		b.WriteString(key + " " + desc + "\n")
	}

	return b.String()
}

// CommonBindings returns common key bindings.
func CommonBindings() []KeyBind {
	return []KeyBind{
		{Key: "q", Description: "quit"},
		{Key: "?", Description: "help"},
	}
}

// NavigationBindings returns navigation key bindings.
func NavigationBindings() []KeyBind {
	return []KeyBind{
		{Key: "j/k", Description: "navigate"},
		{Key: "enter", Description: "select"},
		{Key: "esc", Description: "back"},
	}
}

// SelectionBindings returns selection key bindings.
func SelectionBindings() []KeyBind {
	return []KeyBind{
		{Key: "space", Description: "toggle"},
		{Key: "a", Description: "select all"},
	}
}

// RenderFooter renders a complete footer with bindings.
func RenderFooter(styles tui.Styles, bindings []KeyBind, width int) string {
	kb := NewKeyBinds(styles)
	kb.SetBindings(bindings)
	kb.SetWidth(width)

	content := kb.RenderCompact()

	return lipgloss.NewStyle().
		Width(width).
		Padding(0, 2).
		BorderStyle(lipgloss.NormalBorder()).
		BorderTop(true).
		BorderForeground(lipgloss.Color("#404040")).
		Render(content)
}

File: ash-zig/src/tui/components/keybinds.zig
const std = @import("std");
const ansi = @import("../ansi.zig");
const theme = @import("../theme.zig");

pub const KeyBind = struct {
    key: []const u8,
    description: []const u8,
};

pub const KeyBinds = struct {
    bindings: []const KeyBind = &.{},
    style: theme.Style = theme.Style.init(theme.default_theme),

    pub fn init(bindings: []const KeyBind) KeyBinds {
        return .{ .bindings = bindings };
    }

    pub fn render(self: KeyBinds, writer: anytype) !void {
        for (self.bindings, 0..) |binding, i| {
            if (i > 0) try writer.writeAll("  ");

            try self.style.primary(writer);
            try writer.writeAll(binding.key);
            try self.style.reset(writer);
            try writer.writeAll(":");
            try self.style.muted(writer);
            try writer.writeAll(binding.description);
            try self.style.reset(writer);
        }
    }

    pub fn renderCompact(self: KeyBinds, writer: anytype) !void {
        try self.style.muted(writer);
        for (self.bindings, 0..) |binding, i| {
            if (i > 0) try writer.writeAll(" | ");
            try writer.writeAll(binding.key);
            try writer.writeAll(" ");
            try writer.writeAll(binding.description);
        }
        try self.style.reset(writer);
    }

    pub fn renderVertical(self: KeyBinds, writer: anytype) !void {
        for (self.bindings) |binding| {
            try writer.writeAll("  ");
            try self.style.primary(writer);
            try writer.print("{s: <12}", .{binding.key});
            try self.style.reset(writer);
            try self.style.secondary(writer);
            try writer.writeAll(binding.description);
            try self.style.reset(writer);
            try writer.writeAll("\n");
        }
    }
};

// Common key binding sets
pub const common_bindings = [_]KeyBind{
    .{ .key = "q", .description = "quit" },
    .{ .key = "?", .description = "help" },
};

pub const navigation_bindings = [_]KeyBind{
    .{ .key = "j/k", .description = "navigate" },
    .{ .key = "enter", .description = "select" },
    .{ .key = "esc", .description = "back" },
};

pub const selection_bindings = [_]KeyBind{
    .{ .key = "space", .description = "toggle" },
    .{ .key = "a", .description = "select all" },
};

pub const home_bindings = [_]KeyBind{
    .{ .key = "j/k", .description = "navigate" },
    .{ .key = "enter", .description = "select" },
    .{ .key = "q", .description = "quit" },
};

pub const results_bindings = [_]KeyBind{
    .{ .key = "j/k", .description = "navigate" },
    .{ .key = "space", .description = "toggle" },
    .{ .key = "a", .description = "select all" },
    .{ .key = "enter", .description = "clean" },
    .{ .key = "esc", .description = "back" },
};

pub const confirm_bindings = [_]KeyBind{
    .{ .key = "y", .description = "confirm" },
    .{ .key = "n/esc", .description = "cancel" },
};

pub fn renderFooter(writer: anytype, bindings: []const KeyBind, width: u16, style: theme.Style) !void {
    // Draw separator
    try style.border(writer);
    for (0..width) |_| {
        try writer.writeAll(ansi.box.horizontal);
    }
    try writer.writeAll("\n");

    // Draw bindings
    const kb = KeyBinds.init(bindings);
    kb.style = style;
    try kb.renderCompact(writer);
    try writer.writeAll("\n");
}

// Tests
test "KeyBinds" {
    const bindings = [_]KeyBind{
        .{ .key = "q", .description = "quit" },
        .{ .key = "?", .description = "help" },
    };

    var buf: [256]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buf);

    const kb = KeyBinds.init(&bindings);
    try kb.render(stream.writer());

    const output = stream.getWritten();
    try std.testing.expect(output.len > 0);
}
